### PR TITLE
Better 1.5 bit quantization 

### PR DESCRIPTION
**Pull request by ikawrakow**
While waiting for the [1.58 bit era](https://github.com/ggerganov/llama.cpp/issues/5761#issuecomment-1985097372) to become reality, I decided to see if the current 1.5 quantization in `llama.cpp` can be improved. The answer is yes, and this PR makes the change to `IQ1_S`. It is a breaking change, but I feel this is OK because I don't expect too many 1.5 bpw quants floating around the Internet.

The table shows a PPL comparison between `IQ1_S` on master and this PR. Context is 2048 tokens for LLaMA-v1 and 4096 for all other models. The last column shows the `rms_norm_epsilon` used to generate the PR results.

| Model |  PPL (master) | PPL (PR) | rms_norm_epsilon |
|--:|--:|--:|--:|
| LLaMA-v1-7B   | 17.93 | 14.20 | 5e-5 |
|LLaMA-v1-13B  | 9.619 | 8.941 | 4e-5 |
|LLaMA-v1-30B | 7.564 | 6.999 | 2.5e-5 |
| LLaMA-v2-7B  | 15.33 | 13.51  | 1.875e-5 |
|LLaMA-v2-13B | 9.167 | 8.134 | 2e-5 |
|LLaMA-v2-70B | 5.675 | 5.343 | 3e-5 |
|Mistral-7B         | 10.80 | 11.21 | default |
|Mixtral8x7B      |  7.261 | 6.354 | default |

Apart from Mistral-7B, all results are significantly better. My guess is that we simply got very lucky with Mistral-7B with the previous quantization, considering the unexpectedly large difference between LLaMA-v2-7B and Mistral-7B there.

The new quantization is the exact same size as the one on master, and uses 1.5 bpw (excluding the super-block scale). In the original version these 1.5 bits are spent on a group-of-8 codebook with 512 entries (9 bits), and a 3-bit scale per 8 weights. In the new quantization there are 2048 entries in the codebook (11 bits), along with one 4-bit scale per 32 weights.

### CODE REVIEW

**Comment by Green-Sky** [comment]
> It is a breaking change, but I feel this is OK because I don't expect too many 1.5 bpw quants floating around the Internet.

Shile this is probably true, we still want it to break with a readable message to the user.
I think something like `general.quantization_version` was supposed to be incremented in situations like this. been a while so I don't remember.

**Comment by ikawrakow** [comment]
@ggerganov I see that with the introduction of `ggml-common.h` the i-quant data has been declared `__constant__` on CUDA. Is this based on a thorough comparison versus the original (which does not use `__constant__`)? On my GPU (RTX-4080) using `__constant__` leads to disastrous performance. E.g., for the new `IQ1_S`, I get 17 t/s with `__constant__` vs 204 t/s without. There is only that much constant memory in a GPU, and, when the data does not fit, what happens is up to the gods. I have therefore changed the definition of `GGML_TABLE_BEGIN` to not use `__constant__` on CUDA.

**Comment by ggerganov** [comment]
It's a mistake on my side - I wasn't aware that this can lead to such drastic changes in the performance. Thanks for fixing it

File: ggml-common.h [hunk 1]
```diff
@@ -17,7 +17,10 @@
 #elif defined(GGML_COMMON_IMPL_CUDA)
 #include <cstdint>
 
-#define GGML_TABLE_BEGIN(type, name, size) static const __device__ __constant__ type name[size] = {
+// So, I'm not sure if there are GPU's out there that like having the i-quant data in
+// constant memory. Mine (RTX-4080) definitely does not like it.
+//#define GGML_TABLE_BEGIN(type, name, size) static const __device__ __constant__ type name[size] = {
+#define GGML_TABLE_BEGIN(type, name, size) static const __device__ type name[size] = {
```
Comment (ggerganov) [inline ggml-common.h:23]: I just did tests on RTX 3090, RTX 4090 and A100 and on all of them it is significantly faster to not have the `__constant__` specifier, so it's not just RTX 4080 related

**Review by ggerganov — APPROVED** [review]

**Comment by Artefact2** [comment]
What is the consensus for picking `rms_norm_epsilon`? Spend an hour quantizing, spend another hour calculating PPL, repeat until you have enough values to pick the best?

**Comment by ikawrakow** [comment]
The SYCL code needs to be adjusted to the new quants. As I don't have the ability to test I have not done that, which causes the SYCL tests to fail.

**Pull request merged**